File: lib/services/music_kit_service.dart
import '../core/platform/music_kit_platform.dart';
import '../models/song.dart';
import '../models/album.dart';
import '../models/artist.dart';
import '../models/playlist.dart';
import '../models/music_authorization_status.dart';

class MusicKitService {
  final MusicKitPlatform _platform;

  MusicKitService(this._platform);

  Future<MusicAuthorizationStatus> requestAuthorization() async {
    final result = await _platform.requestAuthorization();
    return MusicAuthorizationStatus.fromString(result);
  }

  Future<List<Song>> getRecentlyPlayed({int limit = 10}) async {
    final data = await _platform.getRecentlyPlayed(limit: limit);
    return data.map((e) => Song.fromJson(e)).toList();
  }

  Future<({List<Song> songs, List<Album> albums, List<Artist> artists})>
      searchCatalog(String term) async {
    final data = await _platform.searchCatalog(term: term);
    final songs = (data['songs'] as List<dynamic>?)
            ?.map((e) => Song.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList() ??
        [];
    final albums = (data['albums'] as List<dynamic>?)
            ?.map((e) => Album.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList() ??
        [];
    final artists = (data['artists'] as List<dynamic>?)
            ?.map((e) => Artist.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList() ??
        [];
    return (songs: songs, albums: albums, artists: artists);
  }

  Future<Artist> getArtist(String id) async {
    final data = await _platform.getArtist(id);
    return Artist.fromJson(data);
  }

  Future<List<Song>> getArtistTopSongs(String id, {int limit = 10}) async {
    final data = await _platform.getArtistTopSongs(id, limit: limit);
    return data.map((e) => Song.fromJson(e)).toList();
  }

  Future<List<Album>> getArtistAlbums(String id) async {
    final data = await _platform.getArtistAlbums(id);
    return data.map((e) => Album.fromJson(e)).toList();
  }

  Future<Album> getAlbum(String id) async {
    final data = await _platform.getAlbum(id);
    return Album.fromJson(data);
  }

  Future<List<Song>> getAlbumTracks(String id) async {
    final data = await _platform.getAlbumTracks(id);
    return data.map((e) => Song.fromJson(e)).toList();
  }

  Future<List<Playlist>> getUserPlaylists() async {
    final data = await _platform.getUserPlaylists();
    return data.map((e) => Playlist.fromJson(e)).toList();
  }

  Future<Playlist> createPlaylist(String name) async {
    final data = await _platform.createPlaylist(name);
    return Playlist.fromJson(data);
  }

  Future<bool> addSongToPlaylist(String songId, String playlistId) async {
    return _platform.addSongToPlaylist(songId, playlistId);
  }

  Future<List<Song>> getPlaylistTracks(String playlistId) async {
    final data = await _platform.getPlaylistTracks(playlistId);
    return data.map((e) => Song.fromJson(e)).toList();
  }

  Future<List<Album>> getUserLibraryAlbums({int limit = 100}) async {
    final data = await _platform.getUserLibraryAlbums(limit: limit);
    return data.map((e) => Album.fromJson(e)).toList();
  }

  Future<List<Song>> getUserLibrarySongs({int limit = 200}) async {
    final data = await _platform.getUserLibrarySongs(limit: limit);
    return data.map((e) => Song.fromJson(e)).toList();
  }

  Future<List<Artist>> getUserLibraryArtists({int limit = 100}) async {
    final data = await _platform.getUserLibraryArtists(limit: limit);
    return data.map((e) => Artist.fromJson(e)).toList();
  }
}

File: ios/Runner/MusicKit/MusicKitService.swift
import Foundation
import MusicKit

class MusicKitService {
    static let shared = MusicKitService()

    func requestAuthorization() async -> String {
        let status = await MusicAuthorization.request()
        switch status {
        case .authorized: return "authorized"
        case .denied: return "denied"
        case .notDetermined: return "notDetermined"
        case .restricted: return "restricted"
        @unknown default: return "unknown"
        }
    }

    func getRecentlyPlayed(limit: Int) async throws -> [[String: Any?]] {
        var request = MusicRecentlyPlayedRequest<Song>()
        request.limit = limit
        let response = try await request.response()
        return MusicItemMapper.toDictArray(response.items)
    }

    func searchCatalog(term: String, types: [String]) async throws -> [String: Any] {
        var request = MusicCatalogSearchRequest(term: term, types: [Song.self, Album.self, Artist.self])
        request.limit = 25
        let response = try await request.response()

        return [
            "songs": MusicItemMapper.toDictArray(response.songs),
            "albums": MusicItemMapper.toDictArray(response.albums),
            "artists": MusicItemMapper.toDictArray(response.artists),
        ]
    }

    func getArtist(id: String) async throws -> [String: Any?] {
        let musicItemID = MusicItemID(id)
        let request = MusicCatalogResourceRequest<Artist>(matching: \.id, equalTo: musicItemID)
        let response = try await request.response()
        guard let artist = response.items.first else {
            throw MusicKitServiceError.notFound
        }
        return artist.toDictionary()
    }

    func getArtistTopSongs(id: String, limit: Int) async throws -> [[String: Any?]] {
        let musicItemID = MusicItemID(id)
        var request = MusicCatalogResourceRequest<Artist>(matching: \.id, equalTo: musicItemID)
        request.properties = [.topSongs]
        let response = try await request.response()
        guard let artist = response.items.first, let topSongs = artist.topSongs else {
            return []
        }
        let limited = MusicItemCollection(Array(topSongs.prefix(limit)))
        return MusicItemMapper.toDictArray(limited)
    }

    func getArtistAlbums(id: String) async throws -> [[String: Any?]] {
        let musicItemID = MusicItemID(id)
        var request = MusicCatalogResourceRequest<Artist>(matching: \.id, equalTo: musicItemID)
        request.properties = [.albums]
        let response = try await request.response()
        guard let artist = response.items.first, let albums = artist.albums else {
            return []
        }
        return MusicItemMapper.toDictArray(albums)
    }

    func getAlbum(id: String) async throws -> [String: Any?] {
        let musicItemID = MusicItemID(id)
        let request = MusicCatalogResourceRequest<Album>(matching: \.id, equalTo: musicItemID)
        let response = try await request.response()
        guard let album = response.items.first else {
            throw MusicKitServiceError.notFound
        }
        return album.toDictionary()
    }

    func getAlbumTracks(id: String) async throws -> [[String: Any?]] {
        let musicItemID = MusicItemID(id)
        var request = MusicCatalogResourceRequest<Album>(matching: \.id, equalTo: musicItemID)
        request.properties = [.tracks]
        let response = try await request.response()
        guard let album = response.items.first, let tracks = album.tracks else {
            return []
        }
        return MusicItemMapper.toDictArray(tracks)
    }

    // MARK: - User Library

    func getUserLibraryAlbums(limit: Int) async throws -> [[String: Any?]] {
        var request = MusicLibraryRequest<Album>()
        request.limit = limit
        let response = try await request.response()
        return MusicItemMapper.toDictArray(response.items)
    }

    func getUserLibrarySongs(limit: Int) async throws -> [[String: Any?]] {
        var request = MusicLibraryRequest<Song>()
        request.limit = limit
        let response = try await request.response()
        return MusicItemMapper.toDictArray(response.items)
    }

    func getUserLibraryArtists(limit: Int) async throws -> [[String: Any?]] {
        var request = MusicLibraryRequest<Artist>()
        request.limit = limit
        let response = try await request.response()
        return MusicItemMapper.toDictArray(response.items)
    }
}

enum MusicKitServiceError: Error {
    case notFound
}
